### PR TITLE
typescript: Add typescript support.

### LINKED ISSUE #518: Add Typescript as a type checker in the codebase
I think we should add the [flow](https://flow.org/en/) to improve the code quality.

### PR DESCRIPTION
**Pull request by vsvipul**
Adds typescript support using tsc for both main and renderer process.

Need to figure out the lint-js part using xo.

**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
Comment (priyank-p) [inline tsconfig.json]: We want to use `esnext` here because we don't want the javascript files to be transpliled for older version, electron always supports newer javascript syntax.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "strict": true,                           /* Enable all strict type-checking options. */
+    "esModuleInterop": true,                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "noImplicitAny": false
```
Comment (priyank-p) [inline tsconfig.json]: This should be `true` to make sure we take full advantage of typescript and let it complain about where we should do type annotations.

File: typings.d.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+declare module '*';
```
Comment (priyank-p) [inline typings.d.ts:5]: Why is this required? The reason I am asking is to know what problem you faced (I know that we need to do something like this to make this work).

File: app/main/menu.ts [hunk 4]
```diff
@@ -211,7 +211,7 @@ class AppMenu {
 	}
 
 	getWindowSubmenu(tabs, activeTabIndex) {
-		const initialSubmenu = [{
+		const initialSubmenu: any[] = [{
```
Comment (priyank-p) [inline app/main/menu.ts:223]: We should declare a interface for this. Using a interface would  will make sure any function consuming this will get exactly what it expected.

(The `noImplicitAny` tsconfig.json option will make sure we can't do this.)

**Review by priyank-p — COMMENTED** [review]
@vsvipul can you split the PR into coherent commits, like each individual commits for: adding dependencies, adding tsconfig.json, migration a file (one commit per file), and for setting up the xo-typescript.

I am looking into why xo-typescript doesn't work!

**Comment by priyank-p** [comment]
@vsvipul I think the xo-typescript thing is not working because you are missing `eslint-config-xo`; the documentation tells us to install that and I think that's what is responsible for adding xo-typescript config to xo. Did not have a chance to test it out yet though :/

**Comment by vsvipul** [comment]
@priyank-p i installed everything mentioned in the documentation.  xo-typescript is working currently but no t with the old js rules, it is using some new rules and some new errors are popping up like these - 
Screenshot from 2019-06-16 10-06-34

File: tsconfig.json (hunk 1 above)
Comment (vsvipul) [reply tsconfig.json]: ok sure!

File: tsconfig.json (hunk 2 above)
Comment (vsvipul) [reply tsconfig.json]: ok gotcha !

File: typings.d.ts (hunk 3 above)
Comment (vsvipul) [reply typings.d.ts:5]: This is to avoid compiler warning for having declaration files for each and every util we use.

File: app/main/menu.ts (hunk 4 above)
Comment (vsvipul) [reply app/main/menu.ts:223]: I changed noImplicitAny to ```true``` but still could do this.

**Comment by vsvipul** [comment]
@priyank-p I have split the commit to multiple commits. Working on migrating other files now.

File: app/main/menu.ts [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
-const { appUpdater } = require('./autoupdater');
+import * as fs from 'fs-extra';
+import * as AdmZip from 'adm-zip';
+import { appUpdater } from './autoupdater';
 
-const ConfigUtil = require(__dirname + '/../renderer/js/utils/config-util.js');
-const DNDUtil = require(__dirname + '/../renderer/js/utils/dnd-util.js');
-const Logger = require(__dirname + '/../renderer/js/utils/logger-util.js');
+import { ConfigUtil } from '../renderer/js/utils/config-util';
+import DNDUtil from '../renderer/js/utils/dnd-util';
```
Comment (akashnimare) [inline app/main/menu.ts]: { DNDUtil } ?

Comment (vsvipul) [reply app/main/menu.ts]: That hasn't been migrated to typescript for now. So, for the sake of testing, left it as it was before. Will change when i migrate DNDUtil to typescript.

File: app/renderer/js/utils/request-util.ts [hunk 6]
```diff
@@ -32,7 +32,7 @@ class RequestUtil {
 		if (certificate) {
 			// To handle case where certificate has been moved from the location in certificates.json
 			try {
-				certificateLocation = fs.readFileSync(certificate);
+				certificateLocation = fs.readFileSync(certificate).toString();
```
Comment (akashnimare) [inline app/renderer/js/utils/request-util.ts]: Is this needed, `toString()`?

**Review by akashnimare — CHANGES_REQUESTED** [review]

File: app/renderer/js/utils/system-util.ts [hunk 7]
```diff
@@ -1,12 +1,17 @@
...
+let instance: any = null;
 
-let instance = null;
+export interface systemUtil {
+	connectivityERR: string[],
+	userAgent: any
+}
 
-class SystemUtil {
+export class systemUtil {
```
Comment (akashnimare) [inline app/renderer/js/utils/system-util.ts]: @priyank-p @vsvipul I don't think export is needed here, can be removed.

File: package.json [hunk 8]
```diff
@@ -142,19 +145,26 @@
     "stylelint": "9.10.1",
     "tap-colorize": "1.2.0",
     "tape": "4.8.0",
-    "xo": "0.18.2"
+    "typescript": "^3.5.2",
+    "xo": "^0.24.0"
   },
   "xo": {
```
Comment (akashnimare) [inline package.json:156]: @priyank-p, @vsvipul do we have an option to put the `xo` config in a separate file since it's getting longer.

File: package.json [hunk 9]
```diff
@@ -180,12 +190,18 @@
...
     ],
     "envs": [
       "node",
       "browser",
       "mocha"
     ]
+  },
+  "dependencies": {
+    "@types/node": "^12.0.8",
+    "@types/requestidlecallback": "^0.1.1"
```
Comment (akashnimare) [inline package.json]: @vsvipul not sure what's this?

**Review by akashnimare — CHANGES_REQUESTED** [review]

File: app/renderer/js/utils/request-util.ts (hunk 6 above)
Comment (priyank-p) [reply app/renderer/js/utils/request-util.ts]: Yes, you need to do that otherwise this would be a buffer. Another way which is think makes more sense is to pass `utf8` as the encoding parameter.

File: package.json (hunk 8 above)
Comment (priyank-p) [reply package.json:156]: No, `xo` doesn't provide a way to use a config file. Though we can use a `.eslintrc` I think.

File: package.json (hunk 9 above)
Comment (priyank-p) [reply package.json]: This was added because @vsvipul was getting a error when he migrated a ts that used this function. I will look into this.

File: app/main/menu.ts (hunk 5 above)
Comment (akashnimare) [reply app/main/menu.ts]: Okay, sounds good.

**Comment by priyank-p** [comment]
@vsvipul I fixed the linting errors we were getting out of nowhere by upgrading `xo` to `v0.22.0`. It is pretty weird that by just upgrading `xo` we would start to get errors we should look into this, I think the eslint upgrade is what's causing this. Note that there are still linting error in the PR, but these are errors we'll want to fix and are coming from eslint-typescript plugin.

I merged the dependencies (into typescript branch) and tsconfig.json commits with some tweaks. I also added a commit which runs `tsc -w` when we run `npm start` still working on `gulp` part of this.

**Comment by priyank-p** [comment]
(Also, I like to mention if you checkout the PR, you will need to do `git clean -d` otherwise `git` won't let you switch branches.)

File: app/renderer/js/utils/request-util.ts [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
 const logger = new Logger({
 	file: `request-util.log`,
 	timestamp: true
 });
 
-let instance = null;
+let instance: any = null;
 
-class RequestUtil {
+export class requestUtil {
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: @vsvipul why did you rename the class? It should be `RequestUtil`

Comment (vsvipul) [reply app/renderer/js/utils/request-util.ts]: We needed to export an instance of RequestUtil to make it work in other files, otherwise tsc was returning an error. So renamed the class and named the instance RequestUtil.

Comment (priyank-p) [reply app/renderer/js/utils/request-util.ts]: We can do that as `export { requestUtil as RequestUtil };`

Comment (vsvipul) [reply app/renderer/js/utils/request-util.ts]: oh, will change. Thanks.

File: app/renderer/js/utils/request-util.ts [hunk 11]
```diff
@@ -23,7 +23,7 @@ class RequestUtil {
 
 	// ignoreCerts parameter helps in fetching server icon and
 	// other server details when user chooses to ignore certificate warnings
-	requestOptions(domain, ignoreCerts) {
+	requestOptions(domain: string, ignoreCerts: boolean) {
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: I added a new interface here for annotating the return value of this function before merging this commit.

File: app/renderer/js/utils/request-util.ts [hunk 12]
```diff
@@ -49,7 +49,7 @@ class RequestUtil {
 		};
 	}
 
-	formatUrl(domain) {
+	formatUrl(domain: string) {
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: Annotated return type here.

File: app/renderer/js/utils/request-util.ts [hunk 13]
```diff
@@ -49,7 +49,7 @@ class RequestUtil {
...
 		const hasPrefix = (domain.indexOf('http') === 0);
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: The linter was complaining about using `startsWith` here so made that change here

File: app/renderer/js/utils/request-util.ts [hunk 14]
```diff
@@ -1,19 +1,19 @@
...
+import { ConfigUtil } from './config-util';
+import { SystemUtil } from './system-util';
 
 const logger = new Logger({
 	file: `request-util.log`,
 	timestamp: true
 });
 
-let instance = null;
+let instance: any = null;
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: change this to `null | SystemUtil`

File: app/renderer/js/utils/system-util.ts [hunk 15]
```diff
@@ -1,12 +1,17 @@
 'use strict';
 
-const {app} = require('electron').remote;
+import { remote } from 'electron';
+const { app } = remote;
+import * as os from 'os';
 
-const os = require('os');
+let instance: any = null;
```
Comment (priyank-p) [inline app/renderer/js/utils/system-util.ts]: changed this to `null | SystemUtil`.

File: app/renderer/js/utils/system-util.ts (hunk 7 above)
Comment (priyank-p) [inline app/renderer/js/utils/system-util.ts]: Removed the interface and directly annoated the class.

Comment (priyank-p) [reply app/renderer/js/utils/system-util.ts]: Fixed this.

File: app/renderer/js/utils/system-util.ts [hunk 16]
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
...
 			} else {
 				return 'Windows 10';
 			}
+		} else {
+			return '';
 		}
 	}
 
-	setUserAgent(webViewUserAgent) {
+	setUserAgent(webViewUserAgent: any) {
```
Comment (priyank-p) [inline app/renderer/js/utils/system-util.ts]: this should be `string`

File: app/renderer/js/utils/system-util.ts [hunk 17]
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
...
+		const platform: string = os.platform();
+		if (platform === 'darwin') {
 			return 'Mac';
-		}
-		if (os.platform() === 'linux') {
+		} else if (platform === 'linux') {
 			return 'Linux';
-		}
-		if (os.platform() === 'win32' || os.platform() === 'win64') {
+		} else if (platform === 'win32' || platform === 'win64') {
```
Comment (priyank-p) [inline app/renderer/js/utils/system-util.ts]: Removed the `win64` check

File: app/renderer/js/utils/system-util.ts [hunk 18]
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
 	}
 
 	getOS() {
-		if (os.platform() === 'darwin') {
+		const platform: string = os.platform();
```
Comment (priyank-p) [inline app/renderer/js/utils/system-util.ts]: I remove the `string` declaration there. The function return a type `NodeJS.Platform` which we want to use instead. Genral rule is you shouldn't annotate type that are implicitly annotated.

For example:
```typescript
// Do
const a: string;
a = 'aa';

// Don't
const a: string = 'aa';
```

File: app/renderer/js/utils/request-util.ts [hunk 19]
```diff
@@ -1,19 +1,19 @@
-const fs = require('fs');
-const Logger = require('./logger-util');
+import * as fs from 'fs';
+import { Logger } from './logger-util';
```
Comment (priyank-p) [inline app/renderer/js/utils/request-util.ts]: This should be just `import Logger from './logger-util'`

**Comment by priyank-p** [comment]
@vsvipul posted the thing I changed when merging the request-util and logger-util commits above.

File: app/renderer/js/utils/system-util.ts (hunk 18 above)
Comment (vsvipul) [reply app/renderer/js/utils/system-util.ts]: Got it! Will keep in mind from now on.

File: app/renderer/js/utils/system-util.ts (hunk 17 above)
Comment (vsvipul) [reply app/renderer/js/utils/system-util.ts]: win64 doesn't exist! So, was remove therefore. See https://nodejs.org/api/os.html#os_os_platform .

**Comment by priyank-p** [comment]
Merged all of the commits with changes to the typescript branch. Thanks @vsvipul for working on this, we just finished migration of the whole `utils` directory which is awesome.

Next, let finish migrating the rest of the three files left in `app/main` directory.

**Comment by abhigyank** [comment]
This is great work @vsvipul @priyank-p 🎉
